File: Program's_Contributed_By_Contributors/Haskell Programs/trappingRainWater.hs
-- PROBLEM: https://leetcode.com/problems/trapping-rain-water/
-- Given n non-negative integers representing an elevation map where the width of each bar is 1, compute how much water it can trap after raining.
-- Author: Nathalia Araújo (@nathaliafab)

-- len > 0: on 1st iteration, arr = p2end, which means len = 0 (ml would get max from an empty list! Ofc we don't want that)
-- ml > p: the max value in the left of p must be higher than p, or else there's no trapped water
-- mr > p: same for the right
trappingRainWater :: [Int] -> [Int] -> Int
trappingRainWater [] _  = 0
trappingRainWater _ [x] = 0
trappingRainWater arr p2end@(p:ps)  | len > 0 && ml > p && mr > p = min ml mr - p + trappingRainWater arr ps
                                    | otherwise                   = trappingRainWater arr ps
	where
		ml  = foldl1 (max) [x | x <- take len arr]
		mr  = foldl1 (max) [x | x <- ps]
		len = length arr - length p2end

main = do
  putStr "Enter array: "
  a <- getLine
  let arr     = read a :: [Int]
  let result  = trappingRainWater arr arr
  print result

File: Program's_Contributed_By_Contributors/Java_Programs/DS Problems/Trapping_rainWater.c
/*
Given an array arr[] of N non-negative integers representing the height of blocks. If width of each block is 1, compute how much water can be trapped between the blocks during the rainy season.
*/
#include <stdio.h>

int max(int x, int y)
{
    if (x > y)
    {
        return x;
    }
    else
    {
        return y;
    }
}
int min(int x, int y)
{
    if (x < y)
    {
        return x;
    }
    else
    {
        return y;
    }
}

int trap(int *height, int heightSize)
{
    int left[heightSize], right[heightSize];
    left[0] = height[0];
    right[heightSize - 1] = height[heightSize - 1];
    for (int i = 1; i < heightSize; i++)
    {
        left[i] = max(left[i - 1], height[i]);
    }
    for (int i = heightSize - 2; i >= 0; i--)
    {
        right[i] = max(right[i + 1], height[i]);
    }
    int ans = 0;
    for (int i = 0; i < heightSize; i++)
    {
        ans += (min(left[i], right[i]) - height[i]);
    }
    return ans;
}
int main()
{
    int n;
    printf("Enter the size--> ");
    scanf("%d", &n);
    printf("Enter the height of the building blocks--> ");
    int height[n];
    for (int i = 0; i < n; i++)
    {
        scanf("%d", &height[i]);
    }
    int ans = trap(height, n);
    printf("\nThe amount of water can be trapped is-->%d\n", ans);
    return 0;
}